src/lang/modifyAst/faces.test.ts

```diff
@@ -0,0 +1,357 @@
+import { err } from '@src/lib/trap'
+import {
+  addShell,
+  retrieveFaceSelectionsFromOpArgs,
+} from '@src/lang/modifyAst/faces'
+import { initPromise } from '@src/lang/wasmUtils'
+import { engineCommandManager, kclManager } from '@src/lib/singletons'
+import { getCodeRefsByArtifactId } from '@src/lang/std/artifactGraph'
+import { createPathToNodeForLastVariable } from '@src/lang/modifyAst'
+import { stringToKclExpression } from '@src/lib/kclHelpers'
+import type { KclCommandValue } from '@src/lib/commandTypes'
+import env from '@src/env'
+
+// Unfortunately, we need the real engine here it seems to get sweep faces populated
+beforeAll(async () => {
+  await initPromise
+  await new Promise((resolve) => {
+    engineCommandManager.start({
+      token: env().VITE_KITTYCAD_API_TOKEN,
+      width: 256,
+      height: 256,
+      setMediaStream: () => {},
+      setIsStreamReady: () => {},
+      callbackOnEngineLiteConnect: () => {
+        resolve(true)
+      },
+    })
+  })
+}, 30_000)
+
+afterAll(() => {
+  engineCommandManager.tearDown()
+})
+
+async function getAstAndArtifactGraph(code: string) {
+  const ast = assertParse(code)
+  await kclManager.executeAst({ ast })
+  const {
+    artifactGraph,
+    execState: { operations },
+  } = kclManager
+  await new Promise((resolve) => setTimeout(resolve, 100))
+  return { ast, artifactGraph, operations }
+}
+
+function createSelectionFromArtifacts(
+  artifacts: Artifact[],
+  artifactGraph: ArtifactGraph
+): Selections {
+  const graphSelections = artifacts.flatMap((artifact) => {
+    const codeRefs = getCodeRefsByArtifactId(artifact.id, artifactGraph)
+    if (!codeRefs || codeRefs.length === 0) {
+      return []
+    }
+
+    return {
+      codeRef: codeRefs[0],
+      artifact,
+    } as Selection
+  })
+  return {
+    graphSelections,
+    otherSelections: [],
+  }
+}
+
+async function getAstAndSolidSelections(code: string) {
+  const { ast, artifactGraph } = await getAstAndArtifactGraph(code)
+  const artifacts = [...artifactGraph.values()].filter((a) => a.type === 'path')
+  if (artifacts.length === 0) {
+    throw new Error('Artifact not found in the graph')
+  }
+  const solids = createSelectionFromArtifacts(artifacts, artifactGraph)
+  return { artifactGraph, ast, solids }
+}
+
+// More complex shell case
+const multiSolids = `size = 100
+case = startSketchOn(XY)
+  |> startProfile(at = [-size, -size])
+  |> line(end = [2 * size, 0])
+  |> line(end = [0, 2 * size])
+  |> tangentialArc(endAbsolute = [-size, size])
+  |> close()
+  |> extrude(length = 65)
+
+thing1 = startSketchOn(case, face = END)
+  |> circle(center = [-size / 2, -size / 2], radius = 25)
+  |> extrude(length = 50)
+
+thing2 = startSketchOn(case, face = END)
+  |> circle(center = [size / 2, -size / 2], radius = 25)
+  |> extrude(length = 50)`
+
+const multiSolidsShell = `${multiSolids}
+shell001 = shell([thing1, thing2], faces = [END, END], thickness = 5)`
```

Comment:
I'm not unable to replicate this behavior with this example code via point-and-click. The problem is that during the Solids step, I cannot select `thing1` and `thing2` separately by clicking, I can only click once and it highlights the whole unified body and writes `thing2` to the KCL.

Reply:
Oh thanks for checking this out. I'll have a look

Reply:
Ok now I remember: in this case we need the sweeps, not bodies, so I was selecting using the feature tree! And that way you can get the flow to work, but you're right, not from scene selection. I'll give it another spin quick to see if I can alleviate you general concern about allowing explicit solid and face selection. Potentially we could keep the solids stuff as optional, only to be used by users that know exactly what to pick otherwise default to infering from face selection